spec/integration/outdated_spec.cr

```diff
@@ -150,4 +148,138 @@ describe "outdated" do
       stdout.should contain("I: Dependencies are up to date!")
     end
   end
+
+  describe "non-release" do
+    describe "without relases" do
+      it "latest any" do
+        with_shard({dependencies: {missing: "*"}}, {missing: "0.1.0+git.commit.#{git_commits("missing").first}"}) do
+          run "shards install"
+
+          stdout = run "shards outdated --no-color"
+          stdout.should contain("I: Dependencies are up to date!")
+        end
+      end
+
+      it "latest branch" do
+        with_shard({dependencies: {missing: {git: git_url("missing"), branch: "master"}}}, {missing: "0.1.0+git.commit.#{git_commits("missing").first}"}) do
+          run "shards install"
+
+          stdout = run "shards outdated --no-color"
+          stdout.should contain("I: Dependencies are up to date!")
+        end
+      end
+
+      it "latest commit" do
+        with_shard({dependencies: {missing: {git: git_url("missing"), commit: git_commits("missing").first}}}, {missing: "0.1.0+git.commit.#{git_commits("missing").first}"}) do
+          run "shards install"
+
+          stdout = run "shards outdated --no-color"
+          stdout.should contain("I: Dependencies are up to date!")
+        end
+      end
+
+      it "outdated any" do
+        commits = git_commits("inprogress")
+        with_shard({dependencies: {inprogress: "*"}}, {inprogress: "0.1.0+git.commit.#{commits[1]}"}) do
+          run "shards install"
+
+          stdout = run "shards outdated --no-color"
+          stdout.should contain("W: Outdated dependencies:")
+          stdout.should contain("  * inprogress (installed: 0.1.0 at #{commits[1][0..6]}, available: 0.1.0 at #{commits.first[0..6]})")
+        end
+      end
+
+      it "outdated branch" do
+        commits = git_commits("inprogress")
+        with_shard({dependencies: {inprogress: {git: git_url("inprogress"), branch: "master"}}}, {inprogress: "0.1.0+git.commit.#{commits[1]}"}) do
+          run "shards install"
+
+          stdout = run "shards outdated --no-color"
+          stdout.should contain("W: Outdated dependencies:")
+          stdout.should contain("  * inprogress (installed: 0.1.0 at #{commits[1][0..6]}, available: 0.1.0 at #{commits.first[0..6]})")
+        end
+      end
+
+      it "outdated commit" do
```

Comment:
There's really no difference to a version restriction: If I pin a version, I want to use that version. If pin a commit, I want to use that commit. When there's a newer version or a newer commit, I want `shards outdated` to tell me about it. I can still choose to keep the pinned version/commit, but I'm aware that newer versions are available.

If we would say a commit restriction is never outdated because the restricted commit itself doesn't update, the same applies to version restrictions really. If I want version 1.0, it's not going to update either, newer releases have different names.

A typical use case for pinning a commit is if you depend on an unreleased feature. It's a more limited restriction than a branch and prevents `shards update` from choosing newer, possibly unstable commits which are untested and could introduce unwanted breaking changes.
In that scenario, I would still like `shards outdated` to tell me if there are newer versions available which I might want to check out. Maybe there's even a new release and I can drop the commit restriction.

Reply:
Ok, it's true that outdated does not force you to update. In that sense I see how you mean is the same as version number.

I rarely use commit restriction. Branches restrictions yes and I value to pick outdated when new commits on that branch are available. I see tagged releases a stable channel, so there is an underlying timeline on tagged releases to me. This timeline is what gives meaning to the outdated here. And I don't translate this timeline to commits for some reason.

But we can see how this plays along, definitely.

Reply:
The behaviour is certainly not ideal yet if the pinned commit is not in the default branch. That's why I left a `TODO` for that.
But once the check is implemented that `pinned_commit` is an ancestor of `HEAD`, this should be more precise.